target/scripts/startup/setup.d/getmail.sh

```diff
@@ -4,38 +4,59 @@ function _setup_getmail() {
+    # Verify correct value for GETMAIL_POLL. Valid are any numbers greater than 0.
+    if ! [[ ${GETMAIL_POLL} =~ ^[0-9]+$ && ${GETMAIL_POLL} -gt 0 ]]; then
+      _log 'warn' "Invalid value for GETMAIL_POLL: ${GETMAIL_POLL}"
+      _log 'warn' "Getmail will be disabled"
+      # return 1
+    fi
 
-    GETMAILRC='/etc/getmailrc.d'
-    CONFIGS=0
+    local GETMAIL_RC ID GETMAIL_DIR
 
-    mkdir -p "${GETMAILRC}"
+    local GETMAIL_CONFIG_DIR='/tmp/docker-mailserver/getmail'
+    local GETMAIL_RC_DIR='/etc/getmailrc.d'
+    local GETMAIL_RC_GENERAL_CF="${GETMAIL_CONFIG_DIR}/getmailrc_general.cf"
+    local GETMAIL_RC_GENERAL='/etc/getmailrc_general'
 
-    # Generate getmailrc configs, starting with the `/etc/getmailrc_general` base config,
-    # Add a unique `message_log` config, then append users own config to the end.
-    for FILE in /tmp/docker-mailserver/getmail-*.cf; do
-      if [[ -f ${FILE} ]]; then
-        CONFIGS=1
-        ID=$(cut -d '-' -f 3 <<< "${FILE}" | cut -d '.' -f 1)
-        local GETMAIL_CONFIG="${GETMAILRC}/getmailrc-${ID}"
+    # Create the directory /etc/getmailrc.d to place the user config in later.
+    mkdir -p "${GETMAIL_RC_DIR}"
 
-        cat /etc/getmailrc_general                                 >"${GETMAIL_CONFIG}"
-        echo -e "message_log = /var/log/mail/getmail-${ID}.log\n" >>"${GETMAIL_CONFIG}"
-        cat "${FILE}"                                             >>"${GETMAIL_CONFIG}"
-      fi
-    done
+    # Check if getmail config directory exists and at least one <ID>.cf file is present.
+    # getmailrc_general.cf is not mandatory and excluded.
+    if ! find "${GETMAIL_CONFIG_DIR}" -type f -name '*.cf' -not -name getmailrc_general.cf 2>/dev/null | grep -q .; then
+      _log 'warn' 'No getmail configration found'
+      _log 'warn' "Getmail will be disabled"
+    fi
 
-    if [[ ${CONFIGS} -eq 1 ]]; then
-      cat >/etc/cron.d/getmail << EOF
-*/${GETMAIL_POLL} * * * * root /usr/local/bin/getmail-cron
-EOF
-      chmod -R 600 "${GETMAILRC}"
+    # Check if custom getmailrc_general.cf file is present.
+    if [[ -f "${GETMAIL_RC_GENERAL_CF}" ]]; then
+      _log 'debug' "Custom 'getmailrc_general.cf' found"
+      cp "${GETMAIL_RC_GENERAL_CF}" "${GETMAIL_RC_GENERAL}"
     fi
 
-    # Both the debug command and cron job (that runs getmail) for getmail
-    # expect this location to exist.
-    GETMAILDIR=/tmp/docker-mailserver/getmail
-    mkdir -p "${GETMAILDIR}"
+    # If no matching filenames are found, and the shell option nullglob is disabled, the word is left unchanged.
+    # If the nullglob option is set, and no matches are found, the word is removed.
+    shopt -s nullglob
+
+    # Generate getmailrc configs, starting with the `/etc/getmailrc_general` base config, then appending users own config to the end.
+    for FILE in "${GETMAIL_CONFIG_DIR}"/*.cf; do
+      if [[ ${FILE} =~ /getmail/(.+)\.cf && ${FILE} != "${GETMAIL_RC_GENERAL_CF}" ]]; then
+        ID=${BASH_REMATCH[1]}
```

Comment:
> Then the ID assignment is just to get a filename without .cf suffix, or the prefix, this is swapping out this previous logic?:

Exactly. It cut off the full path and the file extension. Other solutions invoke external tools + a subshell. Both not that efficient and elegant.

> I assume there is some reason for the extra file pattern beyond BASH_REMATCH, otherwise you could just use basename

Like above. When there is an easy way in bash, it's best practice to avoid calling external tools / subshells.

Reply:
> Other solutions invoke external tools + a subshell. Both not that efficient and elegant.

I've gotta disagree on that reasoning.

```bash
if [[ ${FILE} =~ /getmail/(.+)\.cf &&
  ID=${BASH_REMATCH[1]}
```

vs

```bash
local ID=$(basename -s '.cf' "${FILE}")
```

One relies on capturing a regex group value from a conditional statement into a bash specific variable for that.

My first impression was "where was `BASH_REMATCH` set as a var... does the prefix mean this is a special global..?". So I had to go look that up to confirm where the value was coming from.

Perhaps you could argue the same with someone who is not familiar with the `basename` command, but it's not taking a value from a conditional statement 🤷‍♂️(how is that elegant?)

While it may offer more efficiency I doubt it makes a pragmatic difference here. If we want efficiency we would not be using bash scripts to glue this whole project together.

---

Despite my view on this, if you think your approach is the wiser choice to go with that's fine by me 😎

Reply:
I fully understand your points. BTW: I didn't know about `basename -s`. That is also a perfectly valid and good solution. That avoids something like `basename $FILE | cut -d '.' -f1` with I tried to avoid initially.

There is no right/wrong how to do it. Mostly personal preference I guess. I learned about `BASH_REMATCH` recently and found it pretty elegant (or maybe powerful is a better term) in general. I like regex and I try to avoid the usage of external tools if it makes sense.

An advantage of my solution is, that you and everyone else that didn't know about `BASH_REMATCH` before, will be a bit more be enlightened afterwards 😎

Reply:
> I try to avoid the usage of external tools if it makes sense.
> 
> An advantage of my solution is, that you and everyone else that didn't know about `BASH_REMATCH` before, will be a bit more be enlightened afterwards 😎

I could say the same with `basename` 😝

I don't think it would really be considered as an external tool, it's part of the coreutils package which we use for other common commands. With an advantage that it would be shell agnostic? 🤔 (_not that it matters too much due to reliance on bash in scripts anyway_)

`BASH_REMATCH` does seem a bit interesting, but I'm not sure if I'd ever reach out for it unless it was simpler to grok and use than an equivalent command using `grep` / `sed` / `awk` etc, depending on context. I tend to forget about neat features that I rarely use 😓

Reply:
> I don't think it would really be considered as an external tool

`basename` is a binary and not a bash built-in, therefor calling it an external tool is correct.

It does not matter in our case, just for clarification: Running an external binary in a subshell is expensive, compared to everything build-in directly to bash in regards of the time both need to run. See this example:

```bash
#!/bin/bash

FILE=/tmp/docker-mailserver/getmail/foo.cf

echo "Pure bash:"
time for i in {1..10000}; do
	if [[ ${FILE} =~ /getmail/(.+)\.cf ]]; then
		ID=${BASH_REMATCH[1]}
	fi
done
echo

echo "Using basename":
time for i in {1..10000}; do
	ID=$(basename -s '.cf' $FILE)
done
echo
```

These are the results on my system:

```
Pure bash:

real    0m0.122s
user    0m0.117s
sys     0m0.004s

Using basename:

real    0m14.330s
user    0m10.967s
sys     0m3.777s
```

That's why I try in general to stick with buildins and avoid external tools when not really necessary.

---

> BASH_REMATCH does seem a bit interesting, but I'm not sure if I'd ever reach out for it unless it was simpler to grok and use than an equivalent command using grep / sed / awk etc, depending on context

Exactly, it depends on the context. E.g. if performance is your aim, you would probably go with pure bash, even if more "complicated". On the other hand, I guess you wouldn't stick at all with bash and go with rust or something else 😆 

---

So should I change it, or are you fine with leaving it as it is?

As often, it's just a preference thing and both solutions are perfectly fine here.